File: valtide-api/bff-nestjs/src/otel.ts
/**
 * File: src/otel.ts
 * Author: Sunil+Ai Assistant
 * Date: 2026-09-18
 * Description:
 *   OpenTelemetry bootstrap for the NestJS BFF. Started BEFORE the Nest app
 *   so HTTP + outbound axios calls are auto-instrumented. Exports OTLP/HTTP
 *   to OTEL_EXPORTER_OTLP_ENDPOINT (Grafana LGTM locally; GCP/AWS via the
 *   collector's exporter config). STRICT NO-OP when the endpoint is unset,
 *   so local dev and tests are unaffected. Vendor-neutral by design (the app
 *   only ever speaks OTLP).
 * Source: langchain-upgrade.md §6; observability decision O1
 */
import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { Resource } from '@opentelemetry/resources';
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions';

let sdk: NodeSDK | undefined;

/**
 * Start OpenTelemetry tracing if an OTLP endpoint is configured.
 * @returns True when tracing started; false when it stayed a no-op.
 */
export function startOtel(): boolean {
  if (!process.env.OTEL_EXPORTER_OTLP_ENDPOINT) {
    return false;
  }
  sdk = new NodeSDK({
    resource: new Resource({
      [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME ?? 'valtide-bff',
    }),
    traceExporter: new OTLPTraceExporter(),
    instrumentations: [getNodeAutoInstrumentations()],
  });
  sdk.start();
  // eslint-disable-next-line no-console
  console.log(
    `[otel] BFF tracing enabled -> ${process.env.OTEL_EXPORTER_OTLP_ENDPOINT}`,
  );
  return true;
}

File: valtide-api/agent-fastapi/app/core/otel.py
"""OpenTelemetry API instrumentation (LangChain migration, Phase 6).

File: app/core/otel.py
Author: Sunil+Ai Assistant
Date: 2026-09-18
Description:
    Wires OpenTelemetry tracing for the FastAPI agent service, exporting
    OTLP/HTTP to whatever collector OTEL_EXPORTER_OTLP_ENDPOINT points at
    (Grafana LGTM locally; GCP AMP/Cloud Trace or AWS AMP/X-Ray in cloud,
    selected by the collector's exporter config, not by app code). This is
    the vendor-neutral seam mandated by the observability decision.

    STRICT NO-OP unless OTEL_EXPORTER_OTLP_ENDPOINT is set, so tests, unit
    runs, and the default local stack are entirely unaffected. Distinct
    from LangSmith (agent-graph tracing); this covers the HTTP/API surface.

Source: langchain-upgrade.md §6; observability decision O1
"""
from __future__ import annotations

import logging
import os

logger = logging.getLogger(__name__)


def configure_otel(app, service_name: str) -> bool:
    """Instrument the FastAPI app + outbound httpx with OTLP tracing.

    Args:
        app: The FastAPI application instance to instrument.
        service_name: Logical service name recorded on spans.

    Returns:
        True if tracing was enabled; False when it stayed a no-op (no
        OTEL_EXPORTER_OTLP_ENDPOINT set, or the SDK is unavailable).
    """
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        logger.debug("OTel disabled (no OTEL_EXPORTER_OTLP_ENDPOINT)")
        return False

    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
            OTLPSpanExporter,
        )
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
        from opentelemetry.sdk.resources import SERVICE_NAME, Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
    except ImportError:  # pragma: no cover - optional at runtime
        logger.warning("OpenTelemetry SDK not installed; API tracing stays off")
        return False

    resource = Resource.create({SERVICE_NAME: service_name})
    provider = TracerProvider(resource=resource)
    provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
    trace.set_tracer_provider(provider)

    FastAPIInstrumentor.instrument_app(app)
    HTTPXClientInstrumentor().instrument()
    logger.info("OTel API tracing enabled -> %s (service=%s)", endpoint, service_name)
    return True
